lib/elixir/lib/kernel.ex

```diff
@@ -1567,23 +1575,29 @@ defmodule Kernel do
   defmacro !value
 
   defmacro !{:!, _, [value]} do
-    optimize_boolean(
-      quote do
-        case unquote(value) do
-          x when :"Elixir.Kernel".in(x, [false, nil]) -> false
-          _ -> true
-        end
-      end
-    )
+    build_truthy_check(value, true, false)
   end
 
   defmacro !value do
+    build_truthy_check(value, false, true)
+  end
+
+  defp build_truthy_check(condition, true_body, false_body) do
+    truthy_clause =
+      quote generated: true do
+        value when value != nil and value != false -> unquote(true_body)
```

Comment:
Changed here to follow the same pattern as `cond`, simplifying the optimization in the expansion. I've checked the history, and doesn't seem that there was a particular reason to use `:"Elixir.Kernel".in`

Reply:
So we cannot use `!=` and `and` because otherwise we would need to optimize clauses only after expansion, as we are not quite sure what `and` and `!=` points to. Most likely it points to the kernel implementation but given we can import/unimport them, it may not be true. To avoid the need for expansion, we use the fully qualified module names and calls. So you would have to write: `:erlang.andalso(:erlang.'/='(...), :erlang.'/='(...))`. Given that's too verbose, we use `Elixir.Kernel.in/2` for simplicity.